libafl_frida/src/helper.rs

```diff
@@ -267,31 +267,33 @@ impl FridaInstrumentationHelperBuilder {
         let module_map = Rc::new(ModuleMap::new_with_filter(gum, &mut module_filter));
 
         let mut ranges = RangeMap::new();
+        // Wrap ranges and runtimes in reference-counted refcells in order to move
+        // these references both into the struct that we return and the transformer callback
+        // that we pass to frida-gum.
+        //
+        // These moves MUST occur before the runtimes are init-ed
```

Comment:
Good catch! I'm wondering whether there's any way to enforce this invariant with types.

Reply:
Yeah. It was breaking things for me, very very subtley.

Reply:
I'm open to suggestions for how to enforce this with types.

Reply:
how/why does it make a difference when you create the refcell? It really shouldn't, or we're holding it wrong?

Reply:
I _think_ this is another instance of https://github.com/frida/frida-rust/issues/76. The lifetime of the ranges needs to outlive the lifetime of the transformer. Since ranges would otherwise be returned owned by the helper, their lifetime is tied to the lifetime of the helper. Ideally, we would need to then ensure the helper outlives the transformer. By using `Rc`, we're just manually ensuring that the ranges _can_ outlive the transformer even if the lifetime of the helper ends prematurely.

Reply:
It's not just that. During runtime init, we build assembly sequences which refer to the addresses of their containing runtimes. So it is critical that runtime init happen after the final MOVE. Once we move into the refcell, we never move again.

Reply:
Hmm, that sounds like something that could benefit from `Pin`?

Reply:
Yes, everything that's not pinned is free to be moved at any time, even if it's in a refcell. That it works right now is probably just accidental, but sounds like UB.
Let's Pin inside the refcell? :)

Reply:
I never managed to wrap my head around Pin. @fabianfreyer do you want to take a stab at it?